lib/executor/variable_looping_vus.go

```diff
@@ -187,102 +182,57 @@ func (vlvc VariableLoopingVUsConfig) Validate() []error {
+			steps = append(steps, step)
+			prevScaledVUs = int64(step.PlannedVUs)
+		}
+	}
 
 	for _, stage := range vlvc.Stages {
 		stageEndVUs := stage.Target.Int64
 		stageDuration := time.Duration(stage.Duration.Duration)
-		totalDuration += stageDuration
-
-		stageVUAbsDiff := abs(stageEndVUs - fromVUs)
-		if stageVUAbsDiff == 0 {
-			// We don't have to do anything but update the time offset
-			// if the number of VUs wasn't changed in this stage
-			timeFromStart += stageDuration
-			continue
-		}
 
-		// Handle 0-duration stages, i.e. instant VU jumps
-		if stageDuration == 0 {
-			fromVUs = stageEndVUs
-			prevScaledVUs = es.Scale(stageEndVUs)
-			steps = append(steps, lib.ExecutionStep{
-				TimeOffset: timeFromStart,
-				PlannedVUs: uint64(prevScaledVUs),
-			})
-			continue
-		}
-
-		// For each stage, limit any VU adjustments between the previous
-		// number of VUs and the stage's target to happen at most once
-		// every minIntervalBetweenVUAdjustments. No floats or ratios,
-		// since nanoseconds should be good enough for anyone... :)
-		stepInterval := stageDuration / time.Duration(stageVUAbsDiff)
-		if stepInterval < minIntervalBetweenVUAdjustments {
-			stepInterval = minIntervalBetweenVUAdjustments
-		}
-
-		// Loop through the potential steps, adding an item to the
-		// result only when there's a change in the number of VUs.
-		//
-		// IMPORTANT: we have to be very careful of rounding errors,
-		// both from the step duration and from the VUs. It's especially
-		// important that the scaling via the execution segment should
-		// happen AFTER the rest of the calculations have been done and
-		// we've rounded the global "global" number of VUs.
-		for t := stepInterval; ; t += stepInterval { // Skip the first step, since we've already added that
-			if time.Duration(abs(int64(stageDuration-t))) < minIntervalBetweenVUAdjustments {
-				// Skip the last step of the stage, add it below to correct any minor clock skew
-				break
+		stageVUDiff := stageEndVUs - fromVUs
+		if stageDuration != 0 && stageVUDiff != 0 {
+			var sign int64 = 1
+			if stageVUDiff < 0 {
+				sign = -1
 			}
-			stepGlobalVUs := fromVUs + int64(
-				math.Round((float64(t)*float64(stageEndVUs-fromVUs))/float64(stageDuration)),
-			)
-			stepScaledVus := es.Scale(stepGlobalVUs)
-
-			if stepScaledVus == prevScaledVUs {
-				// only add steps when there's a change in the number of VUs
-				continue
+			// Loop through the potential steps, adding an item to the
+			// result only when there's a change in the number of VUs.
+			for i := sign; i != stageVUDiff; i += sign { // Skip the first step, since we've already added that
+				// VU reservation for gracefully ramping down is handled as a
+				// separate method: reserveVUsForGracefulRampDowns()
+				addStep(lib.ExecutionStep{
+					TimeOffset: timeFromStart + (stageDuration*time.Duration(i))/time.Duration(stageVUDiff),
+					PlannedVUs: uint64(et.ScaleInt64(fromVUs + i)),
```

Comment:
This seems very inefficient... Couldn't the `for` loop above just use a `next()` iterator, like in https://github.com/loadimpact/k6/pull/1285? You don't need to iterate across all potential VU change steps, you already know precisely which VU IDs are apportioned to each segment.

Reply:
I would need to carry it over and so and so forth ... it gets much more complicated, which I intended to do but then this worked and I think this is good enough, for this case, but not for the iterations/s one.

We do this once at the beginning ... the worst case is that we will need to do something like from 0 to 500000 VUs and instead of doing 5000 iterations we would do 500000, once ... 
If you think it is such a big problem I am going to do it but it will get more complicated :D

Reply:
> I would need to carry it over and so and so forth 

I feel like I'm missing something... :confused: True, there would be some complexity surrounding the `for` loop, but there already is, because of the `sign`. Maybe a small lambda or something like that can simplify it, but I'm missing what you mean by "carry it over". 

> the worst case is that we will need to do something like from 0 to 500000 VUs and instead of doing 5000 iterations we would do 500000, once 

I feel like I'm missing something. The current implementation using `ScaleInt64()` has quadratic complexity, since `ScaleInt64()` has a loop inside of it. We don't need `ScaleInt64()` at all, and we can _also_ greatly reduce the `i` values we iterate over.

Reply:
let's say that you go up to 30 and down to 20 and up to 40 again (over some time) and ExecutionTuple is such that you need to take very 7th VU (for the example). 
so I need to emit one start whenever the 7th, 14th,21th,28th, and 35th VU should start if we were executing without segments and so on.

So now I need to iterate over each stage, find that I need to that (in this case, with just every 7th is easier :D), note whether the stage goes up to(or down to) that VU and emit when that happens. I did just try it and couldn't figure it out on the first two tries as it gets complicated around not overshooting it, I think - I will try again with a different approach, as my first two tries got hairy and I couldn't figure out how it should work. 

On `ScaleInt64()`:
1. look at https://github.com/loadimpact/k6/commit/06a231ed4be24858c943e59d1b65a28b9aedfd9e and notice how much faster ScaleInt64 - look at the fact that there are nanoseconds and microseconds in a lot of the places, and that the not preffiled(which is to say not cached) one in a lot of the cases is mostly just the initialization of the ExecutionTuple which happens once. The only bad case is when we have some big numerator as there the loop actually happens fully or at least in big parts. And even then it's also dependant on the number that we are scaling ... 
2. We need it because of the way the current implementation works:
basically it goes over all the usual changes scales the value and checks if it changes given the previous one ... it only changes if we are on an VU that is in the given execution segment this though means that it should wobble or go up and down ...

Reply:
I know that `ScaleInt64` is not that terrible because of the cyclic optimization it uses, but it still is a hammer we probably shouldn't use in a huge loop, when we can both omit it and make the loop smaller.

> So now I need to iterate over each stage, find that I need to that (in this case, with just every 7th is easier :D), note whether the stage goes up to(or down to) that VU and emit when that happens. 

This seems like a good candidate for a small helper method/lambda that you give a `from` and `to` values?